### ISSUE TITLE
Notification is not called on transaction when previous transaction disabled this token.

### ISSUE THREAD

**Issue by molind** [realm/realm-cocoa#4558]
## Goals

Change object.

## Expected Results

Object is changed and notification is called.

## Actual Results

Only object is changed. But notification is not called, because it was disabled during previous commit transaction.

## Steps to Reproduce

1. Commit one transaction without notifying all notification tokens.
2. Commit second transaction, that should notify everyone. 

## Code Sample

```
#import "ViewController.h"
#import <Realm/Realm.h>

@interface ModelA : RLMObject
@property NSString *uuid;
@property NSString *name;
@end

@implementation ModelA
+ (NSString *) primaryKey
{
    return @"uuid";
}
@end

@interface ModelB : RLMObject
@property NSString *uuid;
@property NSString *name;
@end

@implementation ModelB
+ (NSString *) primaryKey
{
    return @"uuid";
}
@end

@implementation ViewController
{
    RLMNotificationToken *_nfA, *_nfB;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    RLMRealm *realm = [RLMRealm defaultRealm];
    
    if([realm isEmpty])
    {
        ModelA *a = [[ModelA alloc] init];
        a.uuid = [[NSUUID UUID] UUIDString];
        
        ModelB *b = [[ModelB alloc] init];
        b.uuid = [[NSUUID UUID] UUIDString];
        
        [realm beginWriteTransaction];
        [realm addObject:a];
        [realm addObject:b];
        [realm commitWriteTransaction];
    }
    
    _nfA = [[ModelA allObjects] addNotificationBlock:^(RLMResults * _Nullable results, RLMCollectionChange * _Nullable change, NSError * _Nullable error) {
        NSLog(@"ModelA chnaged");
    }];
    
    _nfB = [[ModelB allObjects] addNotificationBlock:^(RLMResults * _Nullable results, RLMCollectionChange * _Nullable change, NSError * _Nullable error) {
        NSLog(@"ModelB chnaged");
    }];
    
    [self modifyObjects];
}

-(void) modifyObjects
{
    ModelA *a = [[ModelA allObjects] objectAtIndex:0];
    ModelB *b = [[ModelB allObjects] objectAtIndex:0];
    
    RLMRealm *realm = [RLMRealm defaultRealm];
    [realm beginWriteTransaction];
    a.name = [[NSUUID UUID] UUIDString];
    [realm commitWriteTransactionWithoutNotifying:@[_nfA, _nfB] error:nil]; //Transaction must not trigger _nfB

    [realm beginWriteTransaction];
    b.name = [[NSUUID UUID] UUIDString];
    [realm commitWriteTransaction]; //Transaction is modifying object that must trigger _nfB, but this transaction will not trigger it
    
    [realm beginWriteTransaction];
    b.name = [[NSUUID UUID] UUIDString];
    [realm commitWriteTransaction]; //Next transaction that must trigger _nfB will trigger it ...
    
    [self performSelector:@selector(modifyObjects) withObject:nil afterDelay:2];
}
```
@end

## Version of Realm and Tooling

Realm version: 2.1.2 and 2.2.0

Xcode version: 8.2.1

iOS/OSX version: 10.2

Dependency manager + version: Static framework is added manually

**Comment by istx25** [comment]
Hi @molind. Thank you for reaching out for support. We'll review what you've provided us with and have someone respond as soon as possible with an answer or follow-up questions. We appreciate your patience. :-)

**Comment by tgoyne** [comment]
This will be fixed by https://github.com/realm/realm-object-store/pull/352.

**Comment by istx25** [comment]
Since https://github.com/realm/realm-object-store/pull/352 has been merged, I'm going to close this issue. Thanks all for everyone's contributions.

**Comment by jpsim** [comment]
@istx25 we generally don't close issues until they're fixed in the `master` branch, and since the PR fixing this issue was merged in a _git submodule_ of this repository, we have yet to incorporate it into this project. Reopening until we can advance our version of the object store to include the fix.

**Comment by istx25** [comment]
Ah! Okay, my apologies. I'll know for next time.

**Comment by destman** [comment]
As i understand transaction does not contains information about notifications that must be skipped ?
Does this can lead to bugs on synchronized realms?
I mean what if we commit transaction with commitWriteTransactionWithoutNotifying and flags will be cleared by transaction that comes from other device ...

**Comment by jpsim** [comment]
@destman We're not aware of any bugs regarding skipping notifications other than the one being tracked here and fixed in realm/realm-object-store#352.

**Comment by tgoyne** [comment]
Properly handling transactions coming from other devices was one of the main motivations for building this feature into Realm, as they make it impossible for you to do it yourself. Local tracking of which transaction versions to skip works properly even with sync due to the specific details of how changes from other devices are applied (they obey the local write lock, so if you're in a write transaction no sync changes will be applied on your device until you commit it, and they will then be applied *after* your write transaction).
